signer/src/network/libp2p/bootstrap.rs

```diff
@@ -0,0 +1,314 @@
+        _peer: PeerId,
+        _addr: &Multiaddr,
+        _role_override: libp2p::core::Endpoint,
+        _port_use: libp2p::core::transport::PortUse,
+    ) -> Result<libp2p::swarm::THandler<Self>, libp2p::swarm::ConnectionDenied> {
+        Ok(libp2p::swarm::dummy::ConnectionHandler)
+    }
+
+    fn handle_pending_outbound_connection(
+        &mut self,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _maybe_peer: Option<PeerId>,
+        addresses: &[Multiaddr],
+        effective_role: libp2p::core::Endpoint,
+    ) -> Result<Vec<Multiaddr>, libp2p::swarm::ConnectionDenied> {
+        tracing::debug!(?addresses, "handling pending outbound connection");
+        let addresses = addresses.to_vec();
+
+        // We're only interested in our outbound dialing activity. This _can_ be
+        // a listener if the swarm is attempting a hole-punch.
+        if !effective_role.is_dialer() {
+            return Ok(addresses);
+        }
+
+        for addr in &addresses {
+            if self.is_seed_address(addr) {
+                tracing::debug!(%addr, "attempting to dial seed address");
+            }
+        }
+
+        // Return the addresses as-is.
+        Ok(addresses)
+    }
+
+    fn on_swarm_event(&mut self, event: libp2p::swarm::FromSwarm) {
+        match event {
+            FromSwarm::ConnectionEstablished(e) => {
+                // We're only interested in our outbound activity.
+                if !e.endpoint.is_dialer() {
+                    return;
+                }
+
+                // If we've connected to ourselves, we don't need to do anything.
+                // This might happen if a node has its own address in the seed list.
+                if e.peer_id == self.local_peer_id() {
+                    return;
+                }
+
+                // Increment our total connection count. We're not interested
+                // in counting connections to ourselves.
+                self.connection_count = self.connection_count.saturating_add(1);
+
+                // Get the remote address, and remove the p2p protocol if it exists.
+                let mut addr = e.endpoint.get_remote_address().clone();
+                if matches!(addr.iter().last(), Some(Protocol::P2p(_))) {
+                    addr.pop();
+                }
+
+                // Check if the address is a seed address. If it's not, then
+                // we don't need to do anything so we return early.
+                if self.is_seed_address(&addr) {
+                    tracing::debug!(peer_id = %e.peer_id, %addr, "connected to seed");
+                } else {
+                    return;
+                }
+
+                // Update our connected seeds map.
+                match self.connected_seeds.entry(e.peer_id) {
+                    Entry::Occupied(mut entry) => {
+                        if entry.get_mut().insert(addr) {
+                            tracing::trace!("added connected seed: {:?}", e.peer_id);
```

Comment:
nit: I guess this trace log is also valid for the arm below; we could return if we don't insert here and move the log after the match.

Also, is this tracing log needed? Considering that we already `debug` log (with more info) right above; if we want to keep it, maybe it should also include the `addr` (in case we have multiple addresses for the same peer id); also applied to the same log in `ConnectionClosed` below

Reply:
These are `trace` for debugging purposes, which we generally don't enable (except when I do to debug this). But I added the additional tracing fields since the interpolation needed to be removed anyway. [67c5dc2](https://github.com/stacks-network/sbtc/pull/1257/commits/67c5dc2c13afa7828389e1a9679b0b78f492788c)